ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node.hs

```diff
@@ -657,8 +665,55 @@ protocolInfoCardano protocolParamsByron@ProtocolParamsByron {
     , blockForging
     )
   where
+    CardanoProtocolParams
+      paramsByron
+      paramsShelleyBased
+      paramsShelley
+      paramsAllegra
+      paramsMary
+      paramsAlonzo
+      paramsBabbage
+      paramsConway = paramsCardano
+
+    ProtocolParamsByron {
+          byronGenesis                = genesisByron
+        , byronLeaderCredentials      = mCredsByron
+        , byronMaxTxCapacityOverrides = maxTxCapacityOverridesByron
+        } = paramsByron
+    ProtocolParamsShelleyBased {
+          shelleyBasedGenesis           = genesisShelley
+        , shelleyBasedInitialNonce      = initialNonceShelley
+        , shelleyBasedLeaderCredentials = credssShelleyBased
+        } = paramsShelleyBased
+    ProtocolParamsShelley {
+          shelleyProtVer                = protVerShelley
+        , shelleyMaxTxCapacityOverrides = maxTxCapacityOverridesShelley
+        } = paramsShelley
+    ProtocolParamsAllegra {
+          allegraProtVer                = protVerAllegra
+        , allegraMaxTxCapacityOverrides = maxTxCapacityOverridesAllegra
+        } = paramsAllegra
+    ProtocolParamsMary {
+          maryProtVer                = protVerMary
+        , maryMaxTxCapacityOverrides = maxTxCapacityOverridesMary
+        } = paramsMary
+    ProtocolParamsAlonzo {
+          alonzoProtVer                = protVerAlonzo
+        , alonzoMaxTxCapacityOverrides = maxTxCapacityOverridesAlonzo
+        } = paramsAlonzo
+    ProtocolParamsBabbage {
+          babbageProtVer                = protVerBabbage
+        , babbageMaxTxCapacityOverrides = maxTxCapacityOverridesBabbage
+        } = paramsBabbage
+    ProtocolParamsConway {
+          conwayProtVer                = protVerConway
+        , conwayMaxTxCapacityOverrides = maxTxCapacityOverridesConway
+        } = paramsConway
+
     -- The major protocol version of the last era is the maximum major protocol
     -- version we support.
+    --
+    -- TODO: use index of CardanoProtocolParams NP
```

Comment:
Currently, protocol versions are unique, named fields in each `ProtocolParams` data family instance. Should we introduce a class to extract protocol versions, such that we can do something like extracting the protocol version of the last era in the `NP` of protocol parameters?

Reply:
Yes, versions have been the easiest point of confusion. I think we definitely would benefit from constraining their structure. Like some types to ensure that the version is always increasing, perhaps?

Reply:
In fact, I wonder if the right approach to a HFC block is to start with the details of the first era and then each subsequent era requires a "Transition" argument, which would itself include the protocol params of the new era...

Reply:
Also, see https://github.com/input-output-hk/ouroboros-network/pull/4111#discussion_r1005815258